src/main/java/com/auth0/utils/tokens/SignatureVerifier.java

```diff
@@ -0,0 +1,85 @@
+package com.auth0.utils.tokens;
+
+import com.auth0.exception.IdTokenValidationException;
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.exceptions.JWTDecodeException;
+import com.auth0.jwt.exceptions.JWTVerificationException;
+import com.auth0.jwt.exceptions.SignatureVerificationException;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import com.auth0.jwt.interfaces.JWTVerifier;
+import com.auth0.utils.Asserts;
+
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Represents the verification for an ID Token's signature used when validating an ID token.
+ */
+public abstract class SignatureVerifier {
+
+    private final JWTVerifier verifier;
+    private final List<String> acceptedAlgorithms;
+
+    /**
+     * Get a {@code SignatureVerifier} for use when validating an ID token signed using the HS256 signing algorithm.
+     *
+     * @param secret - the client's secret to use when validating the token's signature.
+     * @return a {@code SignatureVerifier} for use with tokens signed using the HS256 signing algorithm.
+     */
+    public static SignatureVerifier forHS256(String secret) {
+        return new HS256SignatureVerifier(secret);
+    }
+
+    /**
+     * Get a {@code SignatureVerifier} for use when validating an ID token signed using the RS256 signing algorithm.
+     * Callers should provide an implementation of the {@linkplain PublicKeyProvider} to provide the public key used
+     * to verify the ID token's signature.
+     *
+     * @see PublicKeyProvider
+     * @param publicKeyProvider an implementation of {@linkplain PublicKeyProvider} to get the public key.
+     * @return a {@code SignatureVerifier} for use with tokens signed using the RS256 signing algorithm.
+     */
+    public static SignatureVerifier forRS256(PublicKeyProvider publicKeyProvider) {
+        return new RS256SignatureVerifier(publicKeyProvider);
+    }
+
+    /**
+     * Creates a new JWT Signature Verifier.
+     * This instance will validate the token was signed using an expected algorithm
+     * and then proceed to verify its signature
+     *
+     * @param verifier  the instance that knows how to verify the signature.
+     * @param algorithm the accepted algorithms. Must not be null.
+     */
+    SignatureVerifier(JWTVerifier verifier, String... algorithm) {
```

Comment:
if this stays package private, annotate with `@VisibleForTesting` (if available) or just a comment above

Reply:
We'll chat about this today. There may be good reason to make it public so that developers can provide their own verifiers if they wish (including our own usage in mvc-commons).

Reply:
We will keep this package-private and add the `@VisibleForTesting` annotation. For `mvc-commons` we will refactor to use the verification logic in this library in the next major.

Reply:
I updated the comment; it's not exposed for testing purposes, but is used by the implementations.